Visiting current implementation for 2 watch literal scheme

wilberquito opened the pull request:
I am revisiting the code to work on the 2 watch scheme and as I am navigating in the code I find some functions and concepts not very well organized or though...

This PR contains a better way to handled changes of instances and algorithms, as well a better semantic for "updating" the problem. We have a singleton (a store of a problem). You **sync** the problem with the trail or with instance

Closes #439 
Closes #435 
Closes #434 
Closes #432 
Closes #428 
Closes #424 
Closes #423 
Closes #418
Closes #417
Closes #416 
Closes #415 
Closes #414 
Closes #413 
Closes #394 
Closes #378  
Closes #431

wilberquito committed:
Changes of instances and algorithm are better handled

wilberquito committed:
Prettier

wilberquito committed:
Accessing private state by means of method

wilberquito committed:
refactor on occurrence table definition

wilberquito committed:
Import format

wilberquito committed:
renaming function

wilberquito committed:
Uses func to access the var pool

wilberquito committed:
fix: type error for undefined areas in the conflict analysis

wilberquito committed:
fix: decouples logic from learning process and ocurrence table

wilberquito committed:
ref: using minisat naming paradigm

wilberquito committed:
ref: file renamed

wilberquito committed:
feat: watch table entity

wilberquito committed:
Rename type

wilberquito committed:
feat: methods for adding and deleting watches

wilberquito committed:
docs: method execution cost

wilberquito committed:
feat: generalizing the backtrackign definition

wilberquito committed:
fix: var pool tests

wilberquito committed:
ref: uses the same var to repr the working variable

wilberquito committed:
fix: actually throws an error at logging fatal

wilberquito committed:
ref: renaming methods

wilberquito committed:
ref: keep consistency in shared solver implementation

wilberquito committed:
ref: solvers source core

wilberquito committed:
Clarifying some concepts

wilberquito committed:
prettier

wilberquito committed:
fix clause tests

Pafefo commented on src/lib/entities/Clause.svelte.ts:
```diff
@@ -8,24 +9,24 @@ import type { VariablePool } from './VariablePool.svelte.ts';
-	private _learnt: boolean = false;
+	private literals: Literal[] = [];
+	private comments: string[] = $state([]);
+	private cr: number | undefined;
```
CRef type

Pafefo commented on src/lib/entities/ClausePool.svelte.ts:
```diff
@@ -64,26 +65,22 @@ class ClausePool implements IClausePool {
-		}
-		return S;
+	getUnitClauses(): Clause[] {
+		return this.getClauses().filter((c: Clause) => c.isUnit());
```
Better approach!

Pafefo commented on src/lib/entities/ClausePool.svelte.ts:
```diff
@@ -105,31 +102,46 @@ class ClausePool implements IClausePool {
 
-	getLearnt(): Clause[] {
-		return [...this.learnt.values()].map((tag) => this.clauses.get(tag) as Clause);
+	getClausesLearned(): Clause[] {
```
getLearnedClauses() may be a better name?

wilberquito committed:
wip: working on cdcl transition methods

wilberquito committed:
working on cdcl conflict analysis

wilberquito committed:
wip: simplifying logic

wilberquito commented:
@Pafefo we need to talk about the trails list and the `updateClausesToCheck` method

wilberquito committed:
fix: assertive algorithm

wilberquito committed:
prettier

wilberquito committed:
fix: access to unexistent method

wilberquito committed:
fix: import

wilberquito committed:
fix: import

Pafefo committed:
fix import

Pafefo committed:
fix: import

wilberquito committed:
import

wilberquito committed:
Fix merge problems related to events

wilberquito commented:
@Pafefo right now there is a bug related to the propagations.

Pafefo commented:
@wilberquito tomorrow we can check this error

wilberquito committed:
Visiting occurrence list

wilberquito committed:
working on cdcl definiton

wilberquito committed:
moving cdcl solver using occurrence list entity

wilberquito committed:
ref: decouples queue occurrences lists from the cdcl solver

wilberquito committed:
Decoupling conflict analysis from solvers

wilberquito committed:
working on the cdcl conflict analysis

Pafefo commented on src/lib/components/AppComponent.svelte:
```diff
@@ -58,11 +64,13 @@
 	}
 
 	async function stateMachineEvent(s: StateMachineEvent) {
+		console.log('State machine event:', s);
```
Console log

wilberquito replied:
😭 dude

Pafefo commented on src/lib/components/AppComponent.svelte:
```diff
@@ -126,16 +144,23 @@
 		}
 	}
 
+	function init() {
+		onInstanceChanged(getActiveInstance().getInstanceName());
+		onAlgorithmChanged(getConfiguredAlgorithm());
+	}
+
+	init();
```
Okay 👍🏼

Pafefo commented on src/lib/components/assignment/BackjumpingComponent.svelte:
```diff
@@ -31,26 +31,25 @@
+				return getClausePool().at(reason.cRef);
 			} else {
-				logFatal('Reason error', 'The reason is not a backjumping');
+				logFatal('Reason error', 'The reason is not a back-jumping');
```
is "back-jump"  written like this?

Pafefo commented on src/lib/components/tools/OccurrenceListComponent.svelte:
```diff
@@ -24,12 +22,17 @@
 	});
 
-	let checkingIndex: number = $derived(getCheckingIndex());
+	let focusCRef: CRef = $derived(getOccurrenceList().pointedCRef());
```
This is what I said ;)

wilberquito replied:
@Pafefo this is yet to be decided

Pafefo commented on src/lib/components/tools/OccurrenceListComponent.svelte:
```diff
@@ -47,12 +50,12 @@
 <div class="enumerate-clause">
 	<div class="enumerate"></div>
-	<HeadTailComponent inspecting={onPreConflictState}>
+	<HeadTailComponent display={onPreConflictState}>
```
Remember that this should be changed. Now we will always stop in the "all occurrences checked" so this will be always shown

Pafefo commented on src/lib/components/tools/OccurrenceListComponent.svelte:
```diff
@@ -68,15 +71,15 @@
 				</span>
 			</div>
-			<HeadTailComponent inspecting={checkingIndex === index && !onPreConflictState}>
+			<HeadTailComponent display={focusCRef === clause.getCRef() && !onPreConflictState}>
```
Something similar will happen here! Actually, it will never be shown as we will always be in the "pre conflict state" which is the "all occurrences checked"

wilberquito replied:
@Pafefo will be visited soon

Pafefo commented on src/lib/entities/Clause.svelte.ts:
```diff
@@ -34,43 +36,55 @@ class Clause implements Comparable<Clause> {
+		return this.cr as CRef;
+	}
+
+	setCRef(cr: number): void {
```
maybe CRef type?

wilberquito replied:
Nope! only makes sense to add "identifier" to a clause. It is the clause allocator, i.e., clause pool who provides the identifier to a clause when it is added.

Pafefo commented on src/lib/entities/ClausePool.svelte.ts:
```diff
@@ -7,25 +7,26 @@ import {
 class ClausePool implements IClausePool {
 	private clauses: SvelteMap<number, Clause> = new SvelteMap();
-	private learnt: SvelteSet<number> = new SvelteSet();
+	private learnedClauses: SvelteSet<number> = new SvelteSet();
```
CRef type?

Pafefo commented on src/lib/entities/ClausePool.svelte.ts:
```diff
@@ -7,25 +7,26 @@ import {
-	get(clause: number): Clause;
-	getUnitClauses(): SvelteSet<number>;
+	addClause(clause: Clause): CRef;
+	at(clause: number): Clause;
```
Cref type?

Pafefo commented on src/lib/entities/ConflictAnalysis.svelte.ts:
```diff
@@ -0,0 +1,66 @@
+import { logError } from "$lib/states/toasts.svelte.ts";
+import type { Lit } from "$lib/types/types.ts";
+import type { V } from "vitest/dist/chunks/environment.d.cL3nLXbE.js";
```
Is this ever used?

Pafefo commented on src/lib/entities/ConflictAnalysis.svelte.ts:
```diff
@@ -0,0 +1,66 @@
+import { isPropagationReason, type Propagation, type Reason } from "./VariableAssignment.ts";
+import { getClausePool } from "$lib/states/problem.svelte.ts";
+
+export class ConflictAnalysis {
```
I like this class very much!

wilberquito replied:
Nope

Pafefo commented on src/lib/entities/ConflictAnalysis.svelte.ts:
```diff
@@ -0,0 +1,66 @@
+        }
+        this.clause = conflictClause;
+        this.ldlPropagations = ldlPropagations;
+        this.pointer = ldlPropagations.length - 1;
```
Nicely done

Pafefo commented on src/lib/interfaces/Assignment.ts:
```diff
@@ -0,0 +1,6 @@
+
+export interface Assignment {
+	variable: Var;
+	polarity: boolean | undefined;
```
Maybe<boolean>?

wilberquito replied:
@Pafefo this entity has been removed to simplify things

Pafefo commented on src/lib/solvers/shared.svelte.ts:
```diff
@@ -15,119 +16,116 @@ import {
-	} else {
-		trail.push(VariableAssignment.newAutomatedAssignment(variable, algorithm));
-	}
+	const tAssignment: VariableAssignment = manualAssignment
```
What does tAssignment mean?

Pafefo commented on src/lib/entities/WatchTable.svelte.ts:
```diff
@@ -0,0 +1,73 @@
+		this.table = this.makeWatchTable(clauses);
+	}
+
+	deleteWatch(lit: Lit, watch: Watch): void {
```
When a watch is deleted is because of what? Theory concept hehe

wilberquito replied:
A watch must be deleted when the literal is no more watched in the clause. This is done when a new assignment is done (propagation or decision) and the complementary was being watched in the clause, you seek for another watched literal inside the clause.

Pafefo commented on src/lib/states/occurrence-list.svelte.ts:
```diff
@@ -0,0 +1,15 @@
+import { focusOnAssignment, wipeFocusAssignment } from './inspect-assignment.svelte.ts';
+
+let occurrenceList: OccurrenceList = $state(new OccurrenceList());
+
+export function updateOccurrenceList(occurrences: OccurrenceList): void {
+	occurrenceList = occurrences;
+	const literal: Maybe<Lit> = occurrenceList.getLiteral();
+	if (isJust(literal)) focusOnAssignment(fromJust(literal));
+	else wipeFocusAssignment();
+}
```
So here when an occurrence list is selected, the first literal will be automatically set as the observed by the tailhead?

wilberquito replied:
Yes, the idea is that `updateOccurrenceList` is updated by a propagations o first UP. And you seek for the assignment in the trail (if u need to) to focus with the head trail (that might not exist anymore)

wilberquito committed:
feat: simplify conflict analysis

wilberquito committed:
feat: simplify the logic of polarity to hatted

wilberquito committed:
fix: tests

wilberquito committed:
fix: virtual resolution

wilberquito committed:
working on the conflict analysis transitions

wilberquito committed:
wip: cdcl conflict analysis

wilberquito commented:
@Pafefo I've implemented a lot of logic and the balance of lines is negative 😭

wilberquito committed:
feat: virtual resolutions returns either <clause, resolution>

wilberquito committed:
wip: conflict analysis

wilberquito committed:
wip: conflict analysis

Pafefo commented on src/lib/entities/ConflictAnalysis.svelte.ts:
```diff
@@ -1,66 +1,144 @@
+			}
+			const reason: Clause = getClausePool().at((r as Propagation).cRef);
+			const resolvent: Clause = this.clause.resolution(reason);
+			this.updateConflictiveClause(resolvent);
```
Okay now I see that the conflictive clause is updated here

Pafefo commented on src/lib/entities/ConflictAnalysis.svelte.ts:
```diff
@@ -1,66 +1,144 @@
+	}
+
+	private updateConflictiveClause(resolvent: Clause): void {
+		if (!resolvent.falsified()) {
+			logError(
+				'Conflict Analysis Error',
+				'The resolvent clause must be falsified to continue conflict analysis'
```
Nice!

Pafefo commented on src/lib/entities/Literal.svelte.ts:
```diff
@@ -4,63 +4,70 @@ import type { Comparable } from '../interfaces/Comparable.ts';
 export default class Literal implements Comparable<Literal> {
 	private variable: Variable;
-	private polarity: Polarity;
+	private hat: boolean;
```
It is a funny naming, but maybe something like "sign" would be better? True will mean that is positive and false will mean that is negative, thus the literal having the hat.

wilberquito replied:
I like hat

Pafefo committed:
Finished cdcl algorithm

Pafefo committed:
wip: decisions queue

Pafefo committed:
feat: decisions list + trail differ point logic implemented + component fix

Pafefo committed:
wip

Pafefo committed:
cdcl wokrs?

wilberquito committed:
renaming

wilberquito committed:
fix: accessing conflict analysis after cleaned

wilberquito committed:
fix: reset problem instance

The trail editor was not updated with the only one trail

wilberquito committed:
fix: decouples problem reset of reset {new instance, new solver}

wilberquito committed:
fix: pre automode checker

wilberquito committed:
fix: trails wouldn't close after solver trail or solve problem

wilberquito committed:
fix: save position of the previous trails

wilberquito committed:
solver step by step promise

reviewing the step by step logic, simplify the way configured step delay was working

wilberquito committed:
feat: move solver to the next decision and branching

wilberquito committed:
feat: a stronger subscription observable pattern

wilberquito committed:
fix: wipe occurrence list in the share reset

wilberquito commented:
@Pafefo good news! cdcl works! and the undo by reproducing the solver steps also work!

Right now there is a problem at the automatic resolution. The list of trails is not updating in the dom.

wilberquito committed:
fix: rendering trails when the solver machine is in automode

wilberquito committed:
prettier

wilberquito committed:
feat: making smaller the assignments in the trail

wilberquito committed:
wip: adjustment in the assignment preview

wilberquito committed:
feat: working on styling the trail

wilberquito committed:
feat: litener in dom objects for computing heights

Pafefo commented on src/lib/algorithms/backjumping.ts:
```diff
@@ -0,0 +1,23 @@
+		logFatal('Decision level error', 'The entered decision level is not valid');
+	}
+
+	// As the propagations are not meant to be deleted, the DL+1 is obtained
```
Maybe I'll change this because I thought that it was like "As NO propagations are meant to be deleted..."

I would say something like "As the propagations <=DL are not meant to be deleted..."

Pafefo commented on src/lib/components/assignment/BackjumpingComponent.svelte:
```diff
@@ -30,15 +31,21 @@
+			if (isBackJumpingReason(reason)) {
+				return getClausePool().at(reason.cRef);
 			} else {
 				logFatal('Reason error', 'The reason is not a backjumping');
```
I might also change this error message

Pafefo commented on src/lib/components/AppComponent.svelte:
```diff
@@ -1,154 +1,219 @@
-		reloadFromSnapshot(first);
+		wipeDecisions();
+		wipeDifferSequence();
+		// No more occurrence lists queued
+		wipeOccurrenceListQueue();
```
So I understand this will need to be errased, right?

wilberquito replied:
@Pafefo probably this should be moved to the definition of problem...

Pafefo commented on src/lib/components/AppComponent.svelte:
```diff
@@ -1,154 +1,219 @@
-	function lifeCycleController(l: StateMachineLifeCycleEvent): void {
-		const updateAll = () => {
-			updateTrailsEventBus.emit(getTrails());
-			userActionEventBus.emit('record');
-		};
+	function onTrailStacked() {
+		allocDecisionsTrail();
+		stackDifferPos(getLatestTrail().getAssignments().length);
+	}
```
Interesting, I like it. With this the trail does not have to control where it deffers from the rest

Pafefo commented on src/lib/components/AppComponent.svelte:
```diff
@@ -1,154 +1,219 @@
-		subscriptions.push(stateMachineLifeCycleEventBus.subscribe(lifeCycleController));
+		subs.push(
+			solverSignalEventBus
+				.pipe(filter((e) => e === 'finish-step-by-step' || e == 'finish-step'))
```
So, if I don't get this wrong, here there is a pipe that will only let pass those signals that are either a finish step by step of finish step right?

wilberquito replied:
Ye

Pafefo commented on src/lib/entities/ConflictAnalysis.svelte.ts:
```diff
@@ -0,0 +1,144 @@
+export class ConflictAnalysis {
+	clause: Clause;
+	decision: VariableAssignment;
+	ldlPropagations: VariableAssignment[];
```
I would change this to cdlPropagations standing for "Conflictive Decision Level Propagations" thinking about chronological backtracking and so

wilberquito replied:
SGTM

Pafefo commented on src/lib/entities/OccurrenceTable.svelte.ts:
```diff
@@ -0,0 +1,55 @@
+import type Clause from './Clause.svelte.ts';
+import type { CRef, Lit } from '$lib/types/types.ts';
+
+export default class OccurrenceTable {
```
So if I'm not being misled, this class is the place where the occurrences of the literals will appear, right? 

So if I want to know in which clauses the literal -1 appear, here I will know it

wilberquito replied:
Ye

Pafefo committed:
fix: when the problem was unsat and cdcl was runing, and error ocurred

Pafefo committed:
Fix: undo working well

wilberquito committed:
ref: ctrlZ event

wilberquito committed:
fix: merging problems

wilberquito committed:
fix: solver is reinstantiated when instance has changed

this is necessary because some components depends on the state of the solver

wilberquito committed:
Fix UP view

separeted logic in a component
and also fixes the px moving vertically

wilberquito committed:
fix updating trails after single step

wilberquito committed:
fix resolution conflict analysis view

wilberquito committed:
fix resolution conflict analysis view

Pafefo committed:
Merge pull request #419 from udg-lai/trail-styling

feat: making smaller the assignments in the trail

Pafefo committed:
fix: oppacity when inspecting a clause

Pafefo committed:
fix: individual DL works fine

Pafefo committed:
fix: auto step-by-step not updating the delay

wilberquito commented:
@Pafefo I've updated the PR comment adding the issues that this PR closes. Make sure to add any other work related in the issues section and add it to the comment so we can justify our work these days.

Pafefo committed:
wip

Pafefo committed:
fix: Now the cdcl transitions to unsat if empty clause is detected (#429)

Pafefo committed:
Backtracking refactor (#430)

* working on backtracking refactor

* wip

* wip

* fix: Now the cdcl transitions to unsat if empty clause is detected

* wip: backtracking almost working

* wip

* feat: backtacking working perfectly

* feat: backtacking working perfectly

* undo one thing

---------

Co-authored-by: wilberquito <typingwil@gmail.com>

Pafefo committed:
fix: now when reseting the occurrence list is cleaned (#433)

Pafefo committed:
fix: at unary clause shdl is zero (#436)

Pafefo committed:
Dpll reafctor (#434)

* wip

* feat: DPLL working just fiiiiiiine

Pafefo committed:
feat: now the pointer appears on the left in the vertical list ;) (#441)

Pafefo committed:
Fixed the logic of traverse occurrence list (#440)

* feat: Now the logic of the traveseOccurrenceList works perfectly

* fix: await after calling async mehtod

---------

Co-authored-by: wilberquito <typingwil@gmail.com>

Pafefo committed:
Empty clause and unit propagation logic refactor (#438)

* feat: now algorithms retrieve empty and unary(dpll/cdcl only) clauses and inspect them

* feat: now the empty clause is visualitze

* prettier

* A correct way to consult clauses depending on filtering logic

* feat: removed sort as now the coluses will be revised in order when obtained

---------

Co-authored-by: wilberquito <typingwil@gmail.com>

wilberquito committed:
Deletes previous algorithmic undo

Pafefo committed:
Merge pull request #443 from udg-lai/drop-prev-algorithmUndo

Deletes previous algorithmic undo

wilberquito committed:
ref: close context by contextual event

wilberquito committed:
wip: working on multilevel close

Pafefo committed:
Merge pull request #444 from udg-lai/431-close-trail-contextual

ref: close context by contextual event

wilberquito committed:
feat: compute of visible context in the up and resolution views

wilberquito committed:
feat: logic for collapse/expand trail and editor level

wilberquito committed:
feat: collapse all except last decision

wilberquito committed:
fix: control of dl == 0 for preprocessed propagations

wilberquito committed:
fix: open trail context by notification of an event bus

computes correclty the heigh and renames some variables

wilberquito committed:
prettier

wilberquito committed:
merge fixing

wilberquito committed:
align companion trail information at clossing an open dls

wilberquito committed:
fix: compute trails positions after conflict analysis

wilberquito committed:
fix: wipe trails do not notify the starting trail as a stacked

wilberquito committed:
fix: trail editor listen to reseting problem

wilberquito committed:
feat: literals of the up view has background green and color red

this commit also fixes the problem of computing repeated literals on the view
and simplifies previous logic implemented

Pafefo committed:
Merge pull request #448 from udg-lai/fix/revert-up-to

fix: wipe trails do not notify the starting trail as a stacked

Pafefo committed:
Merge pull request #445 from udg-lai/multilevel-close

feat: multilevel dl close

Pafefo commented:
@wilberquito nice work in this last PR. I think this branch is ready to be merged. I'll leave it to do, but maybe better if we do it tomorrow at the lab or through meet to do like a recap

wilberquito commented:
> @wilberquito nice work in this last PR. I think this branch is ready to be merged. I'll leave it to do, but maybe better if we do it tomorrow at the lab or through meet to do like a recap

Sure, maybe not tomorrow though!

wilberquito committed:
fix: minor linter blames

Pafefo reviewed (approved):
I don't like it 👎🏼
